### ISSUE TITLE
Docs

### ISSUE THREAD

**Issue by kieranpotts** [srcflow/srcflow#4]
- [ ] Glossary of terms
- [ ] Beginner's Guide to VC

References:
- http://blog.plasticscm.com/2010/11/version-control-timeline.html
- https://en.wikipedia.org/wiki/Version_control
- https://en.wikipedia.org/wiki/Comparison_of_version-control_software
- https://betterexplained.com/articles/a-visual-guide-to-version-control/
- https://www.smartsheet.com/software-version-control
- https://nvie.com/posts/a-successful-git-branching-model/
- https://www.atlassian.com/git/tutorials/comparing-workflows

**Comment by kieranpotts** [comment]
Some notes from legacy projects:

* The ultimate aim of using a VCS is to facilitate individual developers to do their work in isolation at their own pace, but poor VCS workflows can have the opposite effect and impede this freedom and flexibility. For example, developers might be blocked waiting for commits from another developer, or waiting for merge conflicts on master to be resolved.
* Best practice is to err on the side of continuous integration rather than long-lived feature branches. Integrating all changes regularly helps to avoid big integration issues. If necessary, put in-progress behind feature toggles. Version control systems do not resolve all types of merge issues - you need a formal process.
* Support branches (v1/*) are essential if you need to maintain multiple releases at the same time. Having a single mainline works only if production releases are always incrementing in chronological order. If we need the flexibility of being able to patch prior releases, then production releases don't always increment chronologically: 1.2.0, 1.3.0, 1.2.1, 1.2.2, 1.3.1 etc. Patches for prior minor releases cannot be merged into and tagged on master/next. You can't merge 1.1 into a point after 2.0 on a single branch. Instead, 1.1 will need to exist and be tagged on a branch that is kept separate from master. ... The solution is to branch long-lived support branches. In addition to ``master/next``, we also maintain multiple legacy master branches, one for each prior major release that we continue to support with bug fixes and security patches. These branches are named ``master/<major>.x``, for example ``master/2.x``. Equivalent of Git Flow "support" branches. 
* Pull requests from issue branches should be built automatically and their build status detected before merging into the mainline.
* PRs should be used for everything, from tiny whitespace refactorings to sweeping changes. A healthy team opens many PRs a day.
* If no-one can push directly to master, we need some shared space for pull requested work to live before it gets merged. There are two possible solutions: use development branches in the same shared repository as the masters, or each developer works in their own fork to which only they have access. Both work, but generally forking scales further. The reason is that branches in the shared repo are shared by all team members and contributors, causing everyone to be in the same effective namespace. Easy to push to the wrong place or delete a branch that someone else is working on. These problems are easy to overcome in a small team, but unmanageable when there are hundreds of contributors around the world. ... Only a handful of maintainers need access to the upstream repository and its pristine masters. When a contributor has done some work, they push that work to a branch on their fork and then submit a merge request to the upstream repo.
* The ``master/*`` branches should always progress forward. They should never get reverted to earlier commits. If a bug makes it through to one or more master branches, new commits should be introduced to those branches that fixes the issue.
* All changes should be made available in a master branch as soon as practical, so that over developers can rebase from the latest and build their changes on top of that by continuously updating their working copies. This ensures that we do not end up in integration hell when multiple feature branches need to be merged at once.
* The principles of semantic versioning are integral to our version control workflow. If you are unfamiliar with semantic versioning, please refer to [the documentation](http://semver.org/).

**Comment by kieranpotts** [comment]
A traditional fork-and-branch workflow is used to manage changes. Each change is initiated by opening an issue. If the project maintainers accept a proposed change, someone will be assigned to implement it. External contributors will need to fork the source code repository and undertake the work in their fork. When work on an issue is complete, the contributor will submit a pull request to have their changes merged into the project mainline and queued for future release.

A contributor is any individual who created or comments on an issue or a pull request.... Contributors must submit pull requests (also known as merge requests) from their fork when they want their changes pulled into the upstream repository.

The upstream repository is assigned a small number of maintainers who are responsible for reviewing merge requests, selecting what gets merged in and when, and for tagging releases.

***

Atomic Changes

As a general rule, each contribution should be the work of a single author and be highly focused in its scope, representing a single distinct change. Multiple orthogonal changes should not be mixed together in a single contribution. Doing so makes it harder for the project's maintainers to manage changes and keep track of the state of the project.


***

The ``dev`` branch SHOULD be the default branch of the upstream repository. Pull requests SHOULD be made from ``issue-*`` branches to the ``dev`` branch when changes are ready to be merged back into the mainline. ``issue-*`` branches MAY be deleted after their changes are merged into ``dev`` or released in ``prod``.

***

The ``dev``, ``qa`` and ``prod`` branches are permanent **mainline branches**. These branches MUST NOT be deleted and changes SHOULD NOT be committed directly to them. Instead, temporary **sideline branches** (``issue-*``) SHOULD be used to implement new features and fix bugs, and there SHOULD be a one-to-one relationship between each of these working branches and open issues in the project's issue tracker.

****

Each contribution SHOULD be the work of a single author and be highly focused in its scope, representing a single distinct change. Multiple orthogonal changes SHOULD NOT be mixed together in a single contribution. Doing so makes it harder for the project's maintainers to manage changes and keep track of the state of the project.

Each discrete change to be introduced to the source code MUST have a corresponding issue logged in the upstream repository's issue tracking system.



## Increments

An **increment** is a discrete software update. An increment is typically triggerd by a new issue — commonly a feature request or bug report — being opened in the project's issue tracker. ... Each icrement (or sprint) will have a unique ID that will reference a ticket number in our issue tracking system. The issue number is also represented in feature branch names in our version control system. No matter how small the iterations, each and every change that is started has a one-to-one relationship with a ticket in our issue tracker and a one-to-one relationship with a feature branch in a source code repository.

Issues may be opened by developers or business analysts. Often, the BAs will work in conjunction with the software developers to define the business requirements in a universal language that everyone understands. If the requirements are substantial, the team will try to break down the requirements into smaller, more manageable chunks of work that can each be developed as separate increments — either sequentially or in parallel, or a combination of both.

An increment is typically concluded — and the associated issue closed — by a new software release, which is recorded as a version number bump in the project's changelog.


## Iterations

An increment is made up of one or more iterations. An **iteration** is a single cycle through the following phases:

1. Analyse
2. Experiment
3. Learn

This part of the process is highly adaptive. It is not timeboxed. ... Evolve the methodology to fit the task at hand. ... Iterations are *not* timeboxed. Although some of today's trendiest agile workflow frameworks use timeboxing, we don't find timeboxing useful. When sprints have fixed durations, developers spend a lot of time managing what can and cannot get merged into the project mainlines in time for each release deadline. Developers are also prone to taking shortcuts to squeeze features into scheduled deployments, especially when under pressure from project managers. Timeboxing carries an administrative overhead and, although the intention of timeboxing is to maintain a baseline of quality, it tends to foster a culture that favours short-term quick fixes over discovering more sustainable solutions to problems. The psychological effect of putting deadlines on product development, which is largely a process of research and design, tends to produce low grade goods. ... So, we prefer our sprints (iterations) to be adaptable to the requirements. An iteration may last anything between a couple of hours and a couple of weeks, depending on the scope and complexity of the feature or change under development. But rarely should an iteration take longer than two weeks. An iteration that lasts longer than this is a smell that the initial analysis and scoping phase did not break down the requirements into small enough fragments to be developed incrementally.

We don't distinguish between design and build phases. We design implementations in code, tests, and/or documentation (such as UML). The approach can vary depending on the time at hand. But whatever the approach, the output of the design process is ultimately production-grade code, tests and documentation.

We prioritise quality over velocity. Do as many iterations as necessary to come to the optimum solution that best balances all of the requirements.

**Comment by kieranpotts** [comment]
## Workflow

Git is a very powerful version control system that supports many different workflows. It is hugely important that we have a standardised source control workflow across all of our projects and applications. The following workflow is designed to be simple, scalable, and effective across all of our myriad applications, from native mobile apps to microservices.




### Long-Running Feature Development

Long-running features, and complex features that warrant multiple developers working on them simultaneously, may require multiple branches. In this case you can branch off from the mainline ``test`` branch a temporary ``test/**`` branch that will become the master branch for the feature. Developers will branch their own ``dev/**`` sidelines from the master ``test/**`` branch, and ultimately merge their work back into ``test/**``. Only when the feature is stable, well tested, and ready for integration with the mainline, will the ``test/**`` sideline branch be merged back into the ``test`` mainline.

### Releases

@todo: include notes on tagging releases, including updating of CHANGELOG.

All source repositories have a ``CHANGELOG.md`` file in the root level. It must be kept up-to-date and written to the [Keep a Changelog](http://keepachangelog.com/) standard. Maintaining a changelog makes it easier to see, in human-readable terms, what has changed between releases without having to read through the history of commits in the Git system. And it keeps the revision history separate from the version control system, which may well change in the future.


## Good Practices

Used incorrectly, Git will readily tangle up your code in a web of merge conflicts. Follow these guidelines to keep your Git workflow predictable and efficient.

### Commit Often

When working in your ``dev/**`` brancges, commit small changes, often.

The more time that passes between commits, the greater the likelihood that your changes will trigger merge conflicts with other people's work. Also, commiting often gives you plenty of recent rollback positions for when you make a mistake.

Work in small, testable and deployable increments. Write tests and code simultaneously.

Never use ``git add -A`` or ``git commit -a``. Do not assume that all of your local changes are shippable. Instead use ``git add -p`` or a Git GUI like [SourceTree](https://www.sourcetreeapp.com/) or [SmartGit](http://www.syntevo.com/smartgit/) to help you cherrypick the changes you want to keep and commit. Review your local changes and stage and commit only what you need.

### Meaningful Changeset Messages

Keep you commit messages terse and concise. Most GUIs will truncate them anyway. 

The first line of a commit message is a brief summary of the changeset. The commit summary is mandatory and must be no more than 50 characters. Ideally, it should describe the expected result of the change, rather than the implementation detail of the change.

```
git log --oneline -3

# fca8925 Add Good Practices section to Source Control chapter
# 9630552 Fix a spelling error in README.md
# 5cd2604 Fix invalid h3 markdown formatting
```

Separated from the summary by a blank line is a longer description. This is optional and it mostly comes in handy in merge commits, where ``dev`` branches are pulled into the ``test`` mainline for code review. At this point, the long description is a great place to convey extra information to fellow developers and especially the code reviewers. For example, you may include links to related issues, add notes on possible repercusions of the changeset's merging into the project mainline, and you might even give reasons for your chosen implementation and provide details of other solutions that were considered.

```
git commit

# Merge 'dev/234-fetch-polyfill' into dev
#
# Polyfill required as Fetch API is not supported by all of our target clients:
# http://caniuse.com/#feat=fetch
```

### Merge and Push for Peer Review

Push your commits to your ``dev/**`` branch in the central repository often, for the purpose of backup.

Merge your work into the ``test`` mainline branch, and push it to the central repository, when you have made sufficient progress to justify peer review. But don't wait too long. Large check-ins are less likely than small ones to get reviewed properly.

### Reverse Merge

Rather than integrating dev work into the ``test`` mainline branch, better to do the integration in reverse: first pull the latest commits from ``test`` into your ``dev/**`` branch, and sort out any merge conflicts in the development sideline branch before finally merging everything back into the mainline. 

This workflow helps to keep the mainlines (relatively) pristine. The ``test`` mainline branch may not be perfect all of the time, but the aim should be for the tip of the ``prod`` branch to always be in a deployable state.

To update a ``dev`` branch with the latest work from the ``test`` mainline, use ``git pull --rebase``. Rebasing rewrites the history of the local working branch. It replays your recent local commits on top of the changes that you pull in. This preserves the order of the changes, keeping your work at the tip of the developent branch. Also, pulling without rebasing adds merge commits, which in this context don't serve any useful purpose in the source code's history.

Always rebase from ``test`` immediately before pushing your local commits to a shared central repository. The effect of rebasing is that the push will produce a fast-forward and a nice clean history in the shared repository. The project maintainers will be able to review your work and merge it into the project mainline faster.

By default, pulls are not rebased, but you can make rebasing the default behaviour by modifying Git's global configuration on your local machine:

```
git config --global pull.rebase true
```

### Merge Often

Merge ``dev/**`` branches into mainlines regularly. Integrate new code as soon as it is stable. If necessary, use toggle switches to deactivate imcomplete features.

The sooner that code changes are merged into the mainlines, the sooner other developers can rebase from the mainline, updating their working copies and building their work on top. Frequent merges into the mainline, followed by rebased pulls back to the distributed forks and clones, mitigates integration hell when multiple development branches are integrated into the mainline at once.

When a development branch is merged into the mainline, the merge adds all of the commits in the development branch to the mainline branch. If there are conflicts between the two branches, these will be reconciled with a final merge commit. But when there is no divergent work to merge, Git simply "fast forwards" to the most recent commit from the two branches and does not record an explicit merge commit. But merge commits are useful in this context because they record when specific features were integrated into the project mainline. To force Git to record an explicit merge commit every time, use the "no fast forward" (``--no-ff``) flag with the ``merge`` command. This can be made the default behaviour by running the following command against the central repository:

```
git config --global merge.ff false
```

**Comment by kieranpotts** [comment]
Issues do not need to remain open through the code reviewing, testing and deployment processes. Once changes are integrated, the merge requests and original issues can be closed. But they should not be deleted entirely. For example, testers may wish to refer to the original behavaioural specifications, written in the issue tracker.
